qemu/qemu_s390x.go

```diff
@@ -0,0 +1,121 @@
+//
+//      http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+*/
+
+package qemu
+
+import (
+	"fmt"
+	"os"
+	"strings"
+)
+
+// -pci devices don't play well with Z hence corresponding -ccw devices
+// See https://wiki.qemu.org/Documentation/Platforms/S390X
+const (
+	// NVDIMM is the Non Volatile DIMM device driver.
+	NVDIMM DeviceDriver = "nvdimm"
+
+	// Virtio9P is the 9pfs device driver.
+	Virtio9P DeviceDriver = "virtio-9p-ccw"
+
+	// VirtioNet is the virt-io networking device driver.
+	VirtioNet DeviceDriver = "virtio-net"
+
+	// VirtioNetPCI is the virt-io pci networking device driver.
+	VirtioNetPCI DeviceDriver = "virtio-net-ccw"
+
+	// VirtioSerial is the serial device driver.
+	VirtioSerial DeviceDriver = "virtio-serial-ccw"
+
+	// VirtioBlock is the block device driver.
+	VirtioBlock DeviceDriver = "virtio-blk"
+
+	// Console is the console device driver.
+	Console DeviceDriver = "virtconsole"
+
+	// VirtioSerialPort is the serial port device driver.
+	VirtioSerialPort DeviceDriver = "virtserialport"
+
+	// VHostVSockPCI is the vhost vsock pci driver.
+	VHostVSockPCI DeviceDriver = "vhost-vsock-ccw"
+
+	// Vfio is the vfio driver
+	Vfio DeviceDriver = "vfio-ccw"
+
+	// VirtioSCSI is the virtio-scsi device
+	VirtioSCSI DeviceDriver = "virtio-scsi-ccw"
+)
+
+const (
+	//VhostUserSCSI represents a SCSI vhostuser device type
+	VhostUserSCSI = "vhost-user-scsi-pci"
+
+	//VhostUserNet represents a net vhostuser device type
+	VhostUserNet = "virtio-net-ccw"
+
+	//VhostUserBlk represents a block vhostuser device type
+	VhostUserBlk = "vhost-user-blk-pci"
+)
+
+const (
+	// VhostVSOCKCCW is the VSOCK vhost device type for s390x.
+	VhostVSOCKCCW = "vhost-vsock-ccw"
+)
+
+// QemuParams returns the qemu parameters built out of the VSOCK device.
+func (vsock VSOCKDevice) QemuParams(config *Config) []string {
+	var deviceParams []string
+	var qemuParams []string
+
+	deviceParams = append(deviceParams, fmt.Sprintf("%s", VhostVSOCKCCW))
+
+	if vsock.DisableModern {
+		deviceParams = append(deviceParams, ",disable-modern=true")
+	}
+	if vsock.VHostFD != nil {
+		qemuFDs := config.appendFDs([]*os.File{vsock.VHostFD})
+		deviceParams = append(deviceParams, fmt.Sprintf(",vhostfd=%d", qemuFDs[0]))
+	}
+	deviceParams = append(deviceParams, fmt.Sprintf(",id=%s", vsock.ID))
+	deviceParams = append(deviceParams, fmt.Sprintf(",%s=%d", VSOCKGuestCID, vsock.ContextID))
+
+	qemuParams = append(qemuParams, "-device")
+	qemuParams = append(qemuParams, strings.Join(deviceParams, ""))
+
+	return qemuParams
+}
+
+// QemuDeviceParam converts to the QEMU -device parameter notation
+func (n NetDeviceType) QemuDeviceParam() string {
```

Comment:
well done, now this function is generic and takes the values from `qemu_generic.go` or `qemu_s390x.go` , why don't move it to `qemu.go` ?

Reply:
changed :+1: